File: 기출/KAKAO 2019 Internship/호텔방배정/solution.py
import sys
sys.setrecursionlimit(10**6)
def solution(k, room_number):
    dic = {}
    for num in range(1, k+1):
        dic[num] = num
    
    answer = []
    for number in room_number:
        answer.append(getVacantSpot(dic, number, k))
    return answer
    
def getVacantSpot(dic, number, k):
    if dic[number] == number:
        if number == 10:
            dic[number] = 1
        else:
            dic[number] += 1
        return number
    if number < k:
        return getVacantSpot(dic, number+1, k)
k = 1
tc = [10]
print(solution(k, tc))

File: 기출/KAKAO 2019 Internship/호텔방배정/solution.swift
import Foundation

func find(_ rooms: inout [Int], _ number: Int) -> Int {
    if rooms[number] != number {
        rooms[number] = find(rooms, rooms[number])
    }
    return rooms[number]
}

func union(_ rooms: inout [Int], _ a: Int, _ b: Int) -> Int {
    let aParent = find(rooms, a)
    let bParent = find(rooms, b)
    if aParent <= bParent {
        rooms[bParent] = aParent
    } else {
        rooms[aParetn] = bParent
    }
}

func solution(_ k:Int64, _ room_number:[Int64]) -> [Int64] {
    var answer: [Int] = []
    var rooms = Array(0..<k+1)
    
    for room in room_number {
        let parent = find(rooms, room)
        answer.append(parent)
        if parent < k {
            union(rooms, parent, parent + 1)
        }
    }

    return answer

}